src/interface.ts

```diff
@@ -117,7 +117,7 @@ interface DesktopAgent {
    * await agent.raiseIntent(selectedAction.intent.name, context, selectedApp.name);
    * ```
    */
-  findIntents(intent: String, context?: Context): Promise<Array<ActionMetadata>>;
+  findIntents(intent: string | undefined, context?: Context): Promise<ActionMetadata[]>;
```

Comment:
isn't `intent: string | undefined` equivalent to `intent?: string` - which is a bit more compact and easy to understand? i.e. it is an optional parameter.

Reply:
Although this does result in some unpleasant looking usages ...

```
const context = ...;
// find all intents that relate to the given context
const intents = findIntents(undefined, context);
```

This is a pretty nasty API!

A better alternative might be the following:

```
interface IIntentsQuery {
  intent?: string;
  context?: Context;
}

{
  ...

  findIntents(query: IIntentsQuery): Promise<ActionMetadata[]>;

  ...
}
```

which would be used as follows:

```
const context = ...;
// find all intents that relate to the given context
const intents = findIntents({ context });
```

This makes use of shorthand properties, if you're not familiar with them `{ context }` is equivalent to `{ context: context }`.

Reply:
@ColinEberhardt I agree with all you have said. My PR was to try and make the usage clear of "only context".

The other option is to swap the parameters around for context and intent name, and make the second one optional, e.g.

`const intents= findIntents(context);`

`const intents = findIntents(context, 'StartCall');`

but I am not sure that is better, and it also doesn't deal with `const intents = findIntents('StartCall')`, i.e. without context.

The existing API does not exactly lead you into "the pit of success" for these two use cases.

One alternative is to handle the context-only use case separately and make it nice and explicit, to remove any possibility for confusion, i.e.:

```
findIntent(intent: String, context?: Context): Promise<IntentMetadata>;
findAvailableActions(context: Context): Promise<Array<ActionMetadata>>;
```

Reply:
Maybe we should change the name of the method to 'findAction'?
I'd be in favor of making the intent arg optional.  I don't think there's ambiguity, since the args have different types.  So:

`findActions(intent?: string, context?: Context) : Promise<ActionMetadata[]>`

The Query object approach definitely has merit - but seems a little over-engineered to me unless we foresee other properties we'd add to the query...

Reply:
> That is why I think two separate operations make more sense.

I'll second that :+1:

Reply:
Actually - thinking about this some more, I think the term 'action' is misleading, and could be confused with 'intent'. I'd like to propose an alternative.

Reply:
see #44 for an alternative proposal, which keeps your idea of having two separate methods, and removes action